devel/fortio/logger.go

```diff
@@ -0,0 +1,189 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fortio
+
+import (
+	"flag"
+	"fmt"
+	"log"
+	"runtime"
+	"strings"
+)
+
+// LogLevel is the level of logging (0 Debug -> 6 Fatal).
+type LogLevel int
+
+// Log levels
+const (
+	D LogLevel = iota // Debug
+	V LogLevel = iota // Verbose
+	I LogLevel = iota // Info
+	W LogLevel = iota // Warning
+	E LogLevel = iota // Error
+	C LogLevel = iota // Critical
+	F LogLevel = iota // Fatal
+)
+
+var (
+	level          = I // default is Info and up
+	levelToStrA    []string
+	levelToStrM    map[string]LogLevel
+	logPrefix      = flag.String("logprefix", "> ", "Prefix to log lines before logged messages")
+	logFileAndLine = flag.Bool("logcaller", true, "Logs filename and line number of callers to log")
+)
+
+func init() {
+	levelToStrA = []string{
+		"Debug",
+		"Verbose",
+		"Info",
+		"Warning",
+		"Error",
+		"Critical",
+		"Fatal",
+	}
+	levelToStrM = make(map[string]LogLevel, 2*len(levelToStrA))
+	for l, name := range levelToStrA {
+		// Allow both -loglevel Verbose and -loglevel verbose ...
+		levelToStrM[name] = LogLevel(l)
+		levelToStrM[strings.ToLower(name)] = LogLevel(l)
+	}
+	flag.Var(&level, "loglevel", fmt.Sprintf("loglevel, one of %v", levelToStrA))
+	log.SetFlags(log.Ltime)
+}
+
+// String returns the string representation of the level.
+// Needed for flag Var interface.
+func (l *LogLevel) String() string {
+	return (*l).ToString()
+}
+
+// ToString returns the string representation of the level.
+// (this can't be the same name as the pointer receiver version)
+func (l LogLevel) ToString() string {
+	return levelToStrA[l]
+}
+
+// Set is called by the flags.
+func (l *LogLevel) Set(str string) error {
+	var lvl LogLevel
+	var ok bool
+	if lvl, ok = levelToStrM[str]; !ok {
+		// flag processing already logs the value
+		return fmt.Errorf("should be one of %v", levelToStrA)
+	}
+	SetLogLevel(lvl)
+	return nil
+}
+
+// SetLogLevel sets the log level and returns the previous one.
+func SetLogLevel(lvl LogLevel) LogLevel {
+	prev := level
+	if lvl < D {
+		log.Printf("SetLogLevel called with level %d lower than Debug!", lvl)
+		return -1
+	}
+	if lvl > C {
+		log.Printf("SetLogLevel called with level %d higher than Critical!", lvl)
+		return -1
+	}
+	logPrintf(I, "Log level is now %d %s (was %d %s)\n", lvl, lvl.ToString(), prev, prev.ToString())
+	level = lvl
+	return prev
+}
+
+// GetLogLevel returns the current level
+func GetLogLevel() LogLevel {
+	return level
+}
+
+// LogOn returns true if a given level is currently logged.
+func LogOn(lvl LogLevel) bool {
```

Comment:
my first version was IsLogEnabled() and IsLoggingDebug() but I tried to go "go way" (shortest readable)